경북대 BE_박준희 3단계 - 위시 리스트

jjunii opened the pull request:
# 3단계 위시 리스트

## 구현 기능
 - 위시리스트 기능
    - 상품 추가
    - 상품 목록 조회
    - 상품 삭제
- 인증
    - ArgumentResolver
        - 토큰 유효성 검증
        - 인증된 Member 객체 주입

## 질의 사항
1. WishService에서 상품이 존재하지 않을 시 예외 처리 or 상품 정보 응답을 위해 ProductService를 의존하고 있습니다. 그리고 이에 따라 기존에 private findProductOrThrow()도 접근 제어자를 확장하게 되었는데, 이 설계가 자연스러운지 의문이 들었습니다.
대안으로는 WishService에서 ProductRepository를 직접 사용하는 방법도 생각해보았습니다.
서비스 간 의존과 Repository 직접 접근, 어떤 방식이 더 적절한 설계인지 멘토님의 의견이 궁금합니다.

2. 전체 위시리스트를 반환할 때, 현재는 Wish 테이블에서 productId 목록만 조회한 뒤, 서비스 계층에서 각 productId를 기준으로 Product 정보를 개별 조회하여 ResponseDto로 구성하고 있습니다.
하지만 이 방식은 정보를 개별 조회하는 부분이 비효율적이지 않을까 하는 생각이 들었습니다. 그래서 Repository 계층에서 Wish와 Product를 JOIN하는 쿼리를 사용하고, RowMapper를 활용해서 위시리스트 형태의 DTO를 만들어서 Service 계층으로 넘겨주는 방식도 생각해보았습니다.

    이와 관련해서 현재 저는 JdbcClient 환경에서 굳이 객체 자체를 필드로 둘 필요가 없다고 생각했기 때문에 Wish 엔티티에 id, memberId, productId만 필드로 두고 있는데, JOIN을 사용하는 쿼리로 처리하려면 애초에 Product와 Member 객체 자체를 필드로 두는 것이 도메인 구조상 더 자연스러운지에 대한 고민이 동시에 들었습니다.

    정리하자면, 
    - 서비스 계층에서 반복 조회하는 방식보다 조인 쿼리 + RowMapper를 사용한 DTO 직접 반환 방식이 더 나은지
    - Wish가 ID만 필드로 두는 구조와 연관 객체 자체를 필드로 두는 구조 중 어떤 설계가 더 적절하지
이 두 가지에 대해서 멘토님의 의견이 궁금합니다.

3. 기능이 늘어나면서 세분화된 커스텀 예외를 매번 정의하다 보니 커스텀 예외 클래스 수도 많아지고 GlobalExceptionHandler도 점점 길어지고 있습니다. 공통부분이 있는 예외들은 가능한 하나의 메서드에서 처리하는 식으로 단순화하는 게 좋을지, 아니면 지금처럼 세분화해서 처리하는 게 더 나은지 궁금합니다.

## 추가 구현 예정
- 2단계 피드백 반영
- 테스트코드 구현
- 인증 관련 기능 구체화 및 세분화
    - Filter - 인증 담당
        1. 요청 URL 확인: 회원가입/로그인은 토큰이 필요 없음
        2. Authorization Header 확인: 헤더가 없거나 Bearer 로 시작하지 않으면 401
        3. JWT 유효성 확인: 토큰 검증 후 유효하지 않으면 401
    - Interceptor - 인가 담당
        1. 실행될 컨트롤러에 필요한 역할(role) 확인
        2. 토큰에서 역할 조회 후 일치 여부 확인
        3. 권한이 없으면 403
    - ArgumentResolver - 파라미터 객체 매핑
        1. 토큰에서 사용자 ID 추출
        2. ID로 실제 Member 조회, 없으면 404
        3. 조회한 Member 객체를 컨트롤러 파라미터에 매핑
       
제출 시간이 촉박해서 기본 요구사항만 구현하고 제출드립니다. 추가 구현 예정 사항들도 구현되는대로 변경사항 반영하겠습니다!

jjunii committed:
docs(readme): 구현 목록 추가

jjunii committed:
feat: Wish 관련 엔티티 및 DTO 생성

jjunii committed:
feat: Wish 테이블 설계

jjunii committed:
feat: 위시리스트 상품 추가 기능 구현

jjunii committed:
feat: 위시리스트 상품 목록 조회 기능 구현

jjunii committed:
feat: 위시리스트 상품 삭제 기능 구현

jjunii committed:
feat: 상품 삭제 시 id 검증 구현

jjunii committed:
feat: ID로 사용자 찾기 기능 구현

jjunii committed:
feat: ArgumentResolver 구현 및 토큰 기반 인증 처리

jjunii committed:
fix: 외래키 제약 조건을 고려하여 테스트 DB 초기화 방식 변경

nimunsang commented on src/main/java/gift/controller/WishController.java:
```diff
@@ -0,0 +1,56 @@
+    @PostMapping
+    public ResponseEntity<WishResponseDto> addWish(
+            @RequestBody WishRequestDto wishRequestDto,
+            @LoginMember Member member) {
```
WishController 에서 Member Entity 를 사용하고 있는데요~
ArgumentResolver 로부터 인증된 사용자를 세팅할 때, Dto 로 넘겨주는 방식은 어떨까요? (예를 들면 AuthenticatedMember 처럼요) 
사용자 정보 모두가 필요한 것이 아닌, 인증된 사용자에 대한 id, email 정도만 필요할 것 같아요~
한 번 고민해보세요 :)

nimunsang commented on src/main/java/gift/config/LoginMemberArgumentResolver.java:
```diff
@@ -0,0 +1,44 @@
+
+        String authorization = webRequest.getHeader(HttpHeaders.AUTHORIZATION);
+
+        if (authorization == null || !authorization.startsWith("Bearer ")) {
```
`Bearer ` 라는 문자열 값이 여러 번 사용되는데요~ 
상수로 빼서 사용하신다면 휴먼 에러를 방지할 수 있습니다 :)

nimunsang commented on src/main/java/gift/config/LoginMemberArgumentResolver.java:
```diff
@@ -0,0 +1,44 @@
+            throw new UnAuthenticationException("토큰 형식이 올바르지 않습니다.");
+        }
+
+        String token = authorization.split("Bearer ")[1];
```
split 도 좋지만, 위에서 startsWith 로 검증하셨으니 substring 을 사용하시는 것이 더 명확할 것 같아요!
ex) `authorization.substring(BEARER.length());`

nimunsang commented on src/main/java/gift/repository/WishRepository.java:
```diff
@@ -0,0 +1,16 @@
+
+    void deleteWish(Long wishId);
+
+    Optional<Wish> findWish(Long wishId);
```
현재 메서드명들이 findWish, deleteWish, findAllWishes처럼 사용되는 ID 기준이 명확히 드러나지 않아서,
시그니처를 보기 전까진 memberId 기반인지 wishId 기반인지 파악이 어렵다는 느낌을 받았어요!

물론 IDE에서 파라미터 타입을 보면 알 수는 있지만,
메서드 네이밍만 보고도 충분히 의도를 알 수 있도록 조금 더 명확하게 지어보는 건 어떨까요?

```
List<Wish> findAllWishesByMemberId(Long memberId)

void deleteWishById(Long wishId)

Optional<Wish> findWishById(Long wishId)
```

nimunsang commented on src/main/java/gift/service/WishService.java:
```diff
@@ -0,0 +1,58 @@
+                         return new WishResponseDto(wish.getId(), ProductResponseDto.from(product));
+                     })
+                     .toList();
+    }
```
만약 사용자가 "찜" 한 상품이 100개가 넘어간다면, 
findAllWishes 결과가 100개 나올 것이고, `productService.findProductOrThrow()` 메서드가 100번 실행되어 DB 접근이 1 + 100번 일어나게 되는데요~

`productService.findProductsByIdIn()` 메서드를 생성해서 Repository 에서 id 들에 해당하는 Product 들을 한 방에 꺼내오면 어떨까요~?
한 번 고민해보세요 :)

nimunsang commented on src/main/java/gift/config/LoginMemberArgumentResolver.java:
```diff
@@ -0,0 +1,44 @@
+        String token = authorization.split("Bearer ")[1];
+        Long memberId = jwtProvider.getMemberIdFromToken(token);
+
+        return memberService.getMemberById(memberId);
```
만약 memberId 로 찾은 사용자가 없다면, MemberNotFoundException 이 발생하는데요~
여기서는 "인증된 사용자" 를 원하므로 "인증되지 않은 사용자입니다" 라는 Exception 이 발생하는 것이 의미상 명확할 것 같은데 어떻게 생각하시나요~?

nimunsang commented on src/main/resources/schema.sql:
```diff
@@ -13,4 +13,13 @@ CREATE TABLE member
+    product_id  BIGINT NOT NULL,
+    FOREIGN KEY (member_id) REFERENCES member(id),
+    FOREIGN KEY (product_id) REFERENCES product(id)
 );
```
DDL 잘 작성해주셨네요 :)
조금 더 욕심을 내보자면, 
1. 위시 테이블에서 (member_id, product_id) 에 대한 데이터는 unique 해야 할 것 같아요.
2. 인덱스도 적절하게 걸어보는 것이 어떨까요?
3. 멤버 또는 상품이 삭제되었을 때, wish 도 삭제되어야 할지? 에 대한 고민도 해보셨으면 좋겠습니다 😄

nimunsang reviewed (changes requested):
준희님 안녕하세요 :)
3단계 미션까지 잘 완료해주셨네요 😆 정말 고생 많으셨습니다.
전반적으로 요구사항에 맞게 깔끔하게 코드 작성해주셔서 특별히 리뷰드릴 내용은 없었고, 몇 가지 코멘트 남겨드렸으니 한 번 확인해보세요!
다음 주차부터도 화이팅입니다 🔥🔥

---

질의사항에 대한 답변입니다.

> 1. 서비스 간 의존, Repository 접근 비교
- 아주 좋은 질문이네요 👍 접근 제어자를 확장하신 것을 보고 의도를 이해할 수 있었는데요. (private -> default)
- Service 로 접근하게 된다면, ProductService 의 ProductNotFoundException 과 같은 로직을 ProductService 에서 담당해주므로 WishService 에서는 엄청 편하게 사용할 수 있겠네요. Repository 를 직접 참조하게 되면, 직접 예외처리를 해주어야겠구요. 이러한 관점에서 봤을 때, ProductService 를 직접 의존하는 것은 어색하지 않은 것 같아요 :)

> 2. 전체 위시리스트 반환 시 고민
- 아하, 이미 성능에 대한 고민을 하고 계셨네요 😆
- "서비스 계층에서 반복 조회하는 방식보다 조인 쿼리 + RowMapper를 사용한 DTO 직접 반환 방식이 더 나은지 " 고민에 대해서는, 반복 조회 시에 성능 저하 문제가 발생할 수 있으므로 후자의 방식이 더 좋을 것 같습니다. 아주 좋은 방식인 것 같아요 :)
- "Wish가 ID만 필드로 두는 구조와 연관 객체 자체를 필드로 두는 구조 중 어떤 설계가 더 적절한지" 에 대해서는, 장단점이 있을 것 같은데요~ wish.getProduct().getProductName() 처럼 상품 재조회 없이 객체를 탐색할 수 있는 장점이 있지만, 굳이 필요 없는 상황에서는 조회에 대한 오버헤드가 발생할 것 같아요. (필요없는 필드까지 조회하는 오버헤드) 이 경우, 조회 최적화를 위한 projection이 필요하겠네요. 이렇게 projection 하는 관점에서는 위에 DTO 직접 반환 방식과 같은 맥락인 것 같아요.
- 두 가지 외에 다른 방안으로는, 코멘트에 남겨드린 방법도 있을 것 같습니다 :)

> 3. 커스텀 예외
- 점점 커스텀 예외가 늘어남에 따라 중복코드도 많아지고 유지보수가 어려울 것 같은데요~
- 하나의 방법으로는, CustomException 하나를 정의한 뒤, CustomErrorType 을 Enum 으로 정의하는 방식도 있을 것 같습니다. (ErrorType.PRODUCT_NOT_FOUND .. )
- 사용할 때에는, `throw new CustomException(CustomErrorType.PRODUCT_NOT_FOUND, "상품을 찾을 수 없습니다. 상품 id : {})`  처럼 해두고, ExceptionHandler 는 하나의 동작만을 수행하는거죠. ErrorType 은 필드로 HttpStatus, ErrorCode, ErrorCategory ... 등등 공통으로 에러를 처리하기 위한 필드 를 정의하면 될 것 같습니다. 실제로 현업에서도 이런 방식으로 활용하곤 해요 :) 

---

코드를 먼저 리뷰하며, 이건 개선이 필요하지 않나? 라고 생각했던 부분들을 이미 다 캐치하고 고민하고 계셨군요 👍  많은 고민을 하고 코드를 작성하신 것 같아요 좋습니다 😆
곧바로 Approve 를 해도 괜찮을 것 같지만, 아직 시간이 남았기 때문에 우선 Request Changes 드리고, 피드백 수용할 부분은 고쳐보셔도 좋을 것 같습니다 :)

nimunsang commented:
@jjunii 

준희님은 잘 하시는 것 같으니..
한 가지 추가로 고민해보실 내용을 추가해드리자면,,
현재 인증된 사용자를 ArgumentResolver 를 통해 Controller로부터 사용하고 있는데요~
조금 귀찮지 않나요? wish 를 하려면 사용자에 대한 현재 정보가 필요하니.. controller 로부터 사용자 정보를 받아와야 하니..
wishService 에서 단순히 userId 만 필요한데.. 라는 생각이 들 수 있는데요~ (저만 들 수도 있구요)

wishService 에서 단순히 `CurrentUserContext` 라는 클래스를 주입받아서 사용하는 것은 어떨까요!? (Controller의 ArgumentResolver 는 제거하구요!)

한 번 고민해보셔도 좋을 것 같아요.
그러면 WishService 에서 사용할 때

```
private final CurrentUserContext currentUserContext

Long userId = currentUserContext.getUser();
또는 AuthenticatedUser currentUser = currentUserContext.getAuthenticatedUser();
```
처럼 사용할 수도 있겠네요~

팁) ThreadLocal 에 대해 알아보고, ThreadLocal에 현재 사용자에 대한 정보를 저장하고 꺼내서 사용해보자.

필수는 아니니 "재미로" 시간 여유 되시면 한 번 해보시는 것도 도움되실 것 같습니다 :)

jjunii committed:
refactor: 위시리스트 관련 메서드명 개선

jjunii committed:
refactor: ArgumentResolver 및 인증 처리 로직 개선

- ArgumentResolver가 Entity 대신 DTO를 반환
 - DTO를 WishService까지 전달
- 'Bearer' 토큰 파싱 방식 개선
 - `split()` 대신 상수와 `substring()`을 사용
- 존재하지 않는 사용자에 대한 예외 변경
 - 유효한 토큰으로 존재하지 않는 사용자를 조회 시도 시, 404가 아닌 401로 처리

jjunii committed:
refactor: Wish 테이블 개선

jjunii committed:
refactor: 위시리스트 조회 성능 개선

- 기존 위시리스트 목록 조회 시 발생하는 N+1 문제를 해결하기 위해
'findProductsByIdsIn()' 메서드를 사용하여 위시리스트를 한 번의 쿼리로
조회하도록 변경

jjunii committed:
feat: 위시리스트 중복 상품 추가에 대한 예외 처리 구현

jjunii committed:
feat: JWT에 발급자, 만료 시간 추가

jjunii committed:
refactor: ArgumentResolver에서 인증 로직 Filter로 분리

jjunii committed:
refactor: 인증 정보 처리 방식을 ArgumentResolver에서 ThreadLocal 기반으로 변경

jjunii committed:
feat: 위시리스트 조회 시 추가된 순서로 정렬

jjunii committed:
test: 위시리스트 관련 테스트 코드 작성

jjunii commented:
피드백 주신 내용들은 최대한 반영해보려고 노력했고, 위시리스트 테스트 코드도 작성해보았습니다.

## 추가 질의 사항
1. 필터에서 발생한 예외도 기존 예외들과 일관된 형식의 응답을 유지하기 위해 handlerExceptionResolver를 통해 GlobalExceptionHandler로 넘겨서 처리하고 있습니다. 이 방식이 적절한지, 혹시 더 나은 방법이 있는지 궁금합니다.

2. ArgumentResolver를 제거하고 회원 검증 로직을 Filter로 옮기면서 Filter가 MemberService를 직접 의존하게 됐습니다. 필터가 애플리케이션의 가장 앞단에서 동작하는 컴포넌트인데 서비스 계층을 의존하는 구조가 되어도 괜찮은지 궁금합니다.

---
인증 관련 부분은 거의 처음 해보는 부분이라 깊이 있게 공부하고 설계했다기보다는 자료를 찾아보면서 따라해보며 구현한 부분이 많습니다. 그래서 전체적으로 설계가 올바른지, 또 제가 제대로 이해하고 사용하고 있는지 잘 모르겠습니다..😢 혹시 잘못된 부분이 있다면 많은 피드백 부탁드립니다! 특히 CurrentMemberContext와 Filter 부분이 올바르게 구현됐는지 확인해주시면 감사하겠습니다.

그리고 추가로 피드백 주신 커스텀 예외 처리 방식과 role 기반 권한 처리를 구현하려했는데, 리뷰 일정을 너무 늦추지 않기 위해 해당 부분은 반영하지 못하고 일단 제출드립니다.

nimunsang commented:
@jjunii  

오 제가 추가적으로 제안드린 방법도 구현해보셨군요..!! 정말 고생하셨습니다 😆 
전체적으로 잘 구현해주신 것 같아요. 처음인데 잘 구성해주셨습니다.

---

질문사항 1.
> 필터에서 발생한 예외도 기존 예외들과 일관된 형식의 응답을 유지하기 위해 handlerExceptionResolver를 통해 GlobalExceptionHandler로 넘겨서 처리하고 있습니다. 이 방식이 적절한지, 혹시 더 나은 방법이 있는지 궁금합니다.
- 넵 현재 방식대로 사용하셔도 좋을 것 같습니다~

질문사항 2.
MemberService 를 현재 의존하고 있는 이유는 단순히 "인증된 사용자인지?" 체크를 위함인 것 같습니다.
다만, 사용자의 종류에는 "인증된 사용자" 또는, "그냥 사용자" 도 존재하는데요~ jwt filter 에서 context 를 생성할 때, 굳이 "인증된 사용자" 만을 거를 필요는 없다고 생각합니다.
filter 에서는 단순히 CurrentUserContext 만을 생성해주고, 이 context 에는 memberId, Role 또는 UserType 정도가 들어있으면 좋겠네요. (UserType.ADMIN, UserType.ANONYMOUS 또는 Role.ADMIN, Role.ANONYMOUS, Role.User 등...) 
이렇게 CurrentUserContext 가 생성되었고, 만약 특정 controller 메서드에서 "인증된 사용자" 라는 요구사항이 필요하다면? 별도의 인터셉터를 만들어서 처리할 수 있겠네요. 또는, 컨트롤러 메서드에 어노테이션 기반으로 제어할 수도 있을 것 같아요. (AOP 를 적용해서 공통된 어노테이션을 만들면 좋겠네요~)

그래서 결론적으로는, Filter는 단순히 “누가 요청했는가”에 대한 정보만 컨텍스트에 담고, “이 사용자가 인증되었는가?“는 이후 계층이 판단하는 게 자연스럽습니다. CurrentUserContext 에서 굳이 "인증된 사용자" 를 검증할 필요는 없을 것 같아요. 그러므로, MemberService 를 의존할 필요도 없어진다. 가 결론입니다. Filter는 기술적이며 공통 컴포넌트이기 때문에, 도메인 서비스 계층과의 직접 의존은 구조적 관점에서 좋지 않습니다.

---

전체적으로 잘 구성해주셨어요 :) 고생하셨습니다 😆
다음 주차부터도 화이팅입니다 🔥
